kerneltest/kerneltest.sh

```diff
@@ -0,0 +1,127 @@
+    mv cpu.test kerneltest/
+}
+
+create_initramfs() {
+    echo "creating initramfs"
+    bluebox -e kerneltest/cpu.test
+    mv initramfs.cpio kerneltest
+}
+
+# TODO: host the kernels ourselves to not use cilium's quotas
+download_kernel() {
+    kernel_version=$1
+    echo "downloading kernel $kernel_version"
+    curl -o "kerneltest/kernels/linux-$kernel_version.bz" -s -L -O --fail "https://github.com/cilium/ci-kernels/raw/master/linux-$kernel_version.bz"
+}
+
+use_kernel() {
+    kernel_version=$1
+    if [ ! -f "kerneltest/kernels/linux-$kernel_version.bz" ]; then
+        echo "kernel $kernel_version not found"
+        download_kernel "$kernel_version"
+    fi
+}
+
+github_start() {
+    kernel_version=$1
+    [[ -z "${GITHUB_ACTIONS}" ]] || echo "::group:: running tests on kernel $kernel_version"
+}
+
+github_end() {
+    kernel_version=$1
+    [[ -z "${GITHUB_ACTIONS}" ]] || echo "::endgroup::"
+}
+
+run_tests() {
+    kernel_version=$1
+    memory=$2
+    echo "running tests in qemu"
+    github_start "$kernel_version"
+    # kernel.panic=-1 and -no-reboot ensures we won't get stuck on kernel panic.
+    # TODO: add the ability to not print anything so we can integrate with make and do something like:
+    # vmtests:
+    # 	NO_TTY=1 ./kerneltest/kerneltest.sh
+    qemu-system-x86_64 -no-reboot -append 'printk.devkmsg=on kernel.panic=-1 crashkernel=256M' \
+        -nographic -append "console=ttyS0" -m "$memory" -kernel "kerneltest/kernels/linux-$kernel_version.bz" \
+        -initrd kerneltest/initramfs.cpio | tee "kerneltest/logs/vm_log_$kernel_version.txt"
+    github_end "$kernel_version"
+}
+
+did_test_pass() {
+    kernel_version=$1
+    grep PASS "kerneltest/logs/vm_log_$kernel_version.txt" >/dev/null
+}
+
+check_executable() {
+    executable=$1
+    if ! command -v "$executable" &>/dev/null; then
+        echo "$executable could not be found"
+        exit 1
+    fi
+}
+
+# Initial checks
+check_executable "curl"
+check_executable "qemu-system-x86_64"
+check_executable "bluebox"
+
+# Build artifacts for testing.
+build_bpf
+build_test
+create_initramfs
+
+# Run the tests.
+kernel_versions=("5.10" "5.18" "5.19")
+
+for kernel in "${kernel_versions[@]}"; do
+    use_kernel "$kernel"
+    run_tests "$kernel" "1.5G"
+done
+
+# TODO: return the proper status code
+echo "============="
+echo "Test results:"
+echo "============="
+for kernel in "${kernel_versions[@]}"; do
+    if did_test_pass "$kernel"; then
+        echo "- ✅ $kernel"
+    else
+        echo "- ❌ $kernel"
+    fi
+done
+
+# BUG: we need to check for batch operation support and don't run these tests.
+# use_kernel "5.4"
+# run_tests "5.4" "1.5G"
+# did_test_pass "5.4"
+#
+# BUG: on 4.19. Fix backedge, the verifier spotted the loop and it's not happy, we need to unroll it here.
```

Comment:
This one will take a bit longer to tackle. Will take another look once DWARF unwinding is closer to the finishing line. Either way, I think it would be worth discussing the minimum kernel version we want to support. Let's do that in January. @kakkoyun @v-thakkar

Reply:
Better to run in a dedicated pipeline.

Reply:
Resolving to clear things up here, but let's chat about this in the next few weeks